File: Emulator/Emulated_ICs.py
class ICs:
    def Multiplexer_CD74HCT151E(C: bool = False, B: bool = False, A: bool = False, DATA: list[bool] = [], ENABLE: bool = False) -> list[bool]: 
        dataInputIndex = sum([i[1] for i in [(A, 1), (B, 2), (C, 4)] if i[0]])
        dataInputs = []
        for i in range(7):
            if len(DATA) - 1 >= i: dataInputs.append(DATA[i])
            else: dataInputs.append(False)
        if not ENABLE: return [not (dataInputs[dataInputIndex]), dataInputs[dataInputIndex]] # return [Y = normal, W = inverted]
        return [True, False]
    
    def Demultiplexer_CD74HCT238E(G2: bool = False, nG1: bool = False, nG0: bool = False, C: bool = False, B: bool = False, A: bool= False) -> list[bool]:
        falseList = [False] * 8
        if (not G2) or nG0 or nG1: return falseList
        dataInputIndex = sum([i[1] for i in [(A, 1), (B, 2), (C, 4)] if i[0]])
        falseList[dataInputIndex] = True        
        return falseList
    
    class Shiftregister_SN74LS165AN:
        def __init__(self) -> None:
            self.register = [False] * 8
            self.clk = False
    
        def Pins(self, CLK: bool = False, CLK_INH: bool = False, SH_nLD: bool = False, SER: bool = False, REGISTER_INPUTS: list[bool] = [False] * 8) -> list[bool]:
            if not SH_nLD:  # Load RegisterInputs into register
                self.register = []
                for i in range(8):
                    if len(REGISTER_INPUTS) - 1 >= i:
                        self.register.append(REGISTER_INPUTS[i])
                    else:
                        self.register.append(False)
            else:
                if (CLK and not self.clk) and CLK_INH:  # Rising Edge + CLK not inhibited
                    self.register.insert(0, SER)
                    self.register.pop()
            
            self.clk = CLK
            
            return ([self.register[-1], not self.register[-1]], self.register) # return [Qh = normal, nQh = inverted]

    # Basic 2-INP Logic Gates: 
    def INVERT_CD40106BE(A: bool = False) -> bool: return not A
    def AND_CD4081BE(A: bool = False, B: bool = False) -> bool: return A and B
    def NAND_CD4011BE(A: bool = False, B: bool = False) -> bool: return not (A and B)
    def OR_CD4071BE(A: bool = False, B: bool = False) -> bool: return A or B
    def NOR_CD4001BE(A: bool = False, B: bool = False) -> bool: return not (A or B)
    def XOR_CD4030BE(A: bool = False, B: bool = False) -> bool: return (A and not B) or (not A and B)
    def XNOR_CD4077BE(A: bool = False, B: bool = False) -> bool: return not ((A and not B) or (not A and B))

print(ICs.Multiplexer_CD74HCT151E(C=False, B=False, A=False, DATA=[False, True, True, True, False, True, True, False]))
print(ICs.Demultiplexer_CD74HCT238E(G2=True, C=True, B=True, A=False))

# Shiftregister Code Example:
clock = False
SER_ = True
register1 = ICs.Shiftregister_SN74LS165AN()
for i in range(20):
    clock = not clock
    print(register1.Pins(CLK=clock, CLK_INH=True, SH_nLD=True, SER=SER_))
    SER_ = False

File: Emulator/Emulated_ICs.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

void Multiplexer_CD74HCT151E(bool C, bool B, bool A, bool DATA[], int data_length, bool ENABLE, bool output[2]) {
    int dataInputIndex = (A * 1) + (B * 2) + (C * 4);
    bool dataInputs[8] = {false};

    for (int i = 0; i < 8; i++) {
        if (i < data_length) {
            dataInputs[i] = DATA[i];
        } else {
            dataInputs[i] = false;
        }
    }

    if (!ENABLE) {
        output[0] = dataInputs[dataInputIndex];
        output[1] = !dataInputs[dataInputIndex]; // return [Y = normal, W = inverted]
    } else {
        output[0] = true;
        output[1] = false;
    }
}

void Demultiplexer_CD74HCT238E(bool G2, bool nG1, bool nG0, bool C, bool B, bool A, bool output[8]) {
    bool falseList[8] = {false};

    if (!G2 || nG0 || nG1) {
        memcpy(output, falseList, 8 * sizeof(bool));
        return;
    }

    int dataInputIndex = (A * 1) + (B * 2) + (C * 4);
    falseList[dataInputIndex] = true;
    memcpy(output, falseList, 8 * sizeof(bool));
}

// Shiftregister_SN74LS165AN
typedef struct {
    bool register_[8];
    bool clk;
} Shiftregister_SN74LS165AN;

void Shiftregister_SN74LS165AN_init(Shiftregister_SN74LS165AN *reg) {
    memset(reg->register_, false, 8 * sizeof(bool));
    reg->clk = false;
}

void Shiftregister_SN74LS165AN_Pins(Shiftregister_SN74LS165AN *reg, bool CLK, bool CLK_INH, bool SH_nLD, bool SER, bool REGISTER_INPUTS[], int reg_input_length, bool output[2]) {
    if (!SH_nLD) { // Load RegisterInputs into register
        for (int i = 0; i < 8; i++) {
            if (i < reg_input_length) {
                reg->register_[i] = REGISTER_INPUTS[i];
            } else {
                reg->register_[i] = false;
            }
        }
    } else {
        if ((CLK && !reg->clk) && CLK_INH) { // Rising Edge + CLK not inhibited
            memmove(&reg->register_[1], &reg->register_[0], 7 * sizeof(bool));
            reg->register_[0] = SER;
        }
    }

    reg->clk = CLK;
    output[0] = reg->register_[7]; // return [Qh = normal, nQh = inverted]
    output[1] = !reg->register_[7];
}

// Basic 2-INP Logic Gates
bool INVERT_CD40106BE(bool A) { return !A; }
bool AND_CD4081BE(bool A, bool B) { return A && B; }
bool NAND_CD4011BE(bool A, bool B) { return !(A && B); }
bool OR_CD4071BE(bool A, bool B) { return A || B; }
bool NOR_CD4001BE(bool A, bool B) { return !(A || B); }
bool XOR_CD4030BE(bool A, bool B) { return (A && !B) || (!A && B); }
bool XNOR_CD4077BE(bool A, bool B) { return !((A && !B) || (!A && B)); }

int main() {
    Shiftregister_SN74LS165AN regA;
    Shiftregister_SN74LS165AN_init(&regA);

    bool regA_Qh_output[2];
    bool regA_REGISTER_INPUTS[8] = {true, true, true, true, true, true, true, true};
    Shiftregister_SN74LS165AN_Pins(&regA, false, false, false, false, regA_REGISTER_INPUTS, 8, regA_Qh_output);

    bool OUTPUT[2];
    bool DATA[1] = {regA_Qh_output[0]};
    Multiplexer_CD74HCT151E(false, false, false, DATA, 1, false, OUTPUT);

    printf("%d\n", regA_Qh_output[0]); // Print regA_Qh
    printf("%d\n", OUTPUT[0]); // Print OUTPUT

    return 0;
}
